src/network_analysis.jl

```diff
@@ -0,0 +1,702 @@
+    m, n = size(incidencemat)
+    graph = Graphs.DiGraph(m)
+    rows = rowvals(incidencemat)
+    vals = nonzeros(incidencemat)
+    for j in 1:n
+        inds = nzrange(incidencemat, j)
+        row = rows[inds]
+        val = vals[inds]
+        if val[1] == -1
+            Graphs.add_edge!(graph, row[1], row[2])
+        else
+            Graphs.add_edge!(graph, row[2], row[1])
+        end
+    end
+    return graph
+end
+
+
+### Linkage, Deficiency, Reversibility ###
+
+"""
+    linkageclasses(rn::ReactionSystem)
+
+Given the incidence graph of a reaction network, return a vector of the
+connected components of the graph (i.e. sub-groups of reaction complexes that
+are connected in the incidence graph).
+
+Notes:
+- Requires the `incidencemat` to already be cached in `rn` by a previous call to
+  `reactioncomplexes`.
+
+For example,
+```julia
+sir = @reaction_network SIR begin
+    β, S + I --> 2I
+    ν, I --> R
+end
+complexes,incidencemat = reactioncomplexes(sir)
+linkageclasses(sir)
+```
+gives
+```julia
+2-element Vector{Vector{Int64}}:
+ [1, 2]
+ [3, 4]
+```
+"""
+function linkageclasses(rn::ReactionSystem)
+    nps = get_networkproperties(rn)
+    if isempty(nps.linkageclasses)
+        nps.linkageclasses = linkageclasses(incidencematgraph(rn))
+    end
+    nps.linkageclasses
+end
+
+linkageclasses(incidencegraph) = Graphs.connected_components(incidencegraph)
+
+@doc raw"""
+    deficiency(rn::ReactionSystem)
+
+Calculate the deficiency of a reaction network.
+
+Here the deficiency, ``\delta``, of a network with ``n`` reaction complexes,
+``\ell`` linkage classes and a rank ``s`` stoichiometric matrix is
+
+```math
+\delta = n - \ell - s
+```
+
+Notes:
+- Requires the `incidencemat` to already be cached in `rn` by a previous call to
+  `reactioncomplexes`.
+
+For example,
+```julia
+sir = @reaction_network SIR begin
+    β, S + I --> 2I
+    ν, I --> R
+end
+rcs,incidencemat = reactioncomplexes(sir)
+δ = deficiency(sir)
+```
+"""
+function deficiency(rn::ReactionSystem)
+    nps = get_networkproperties(rn)
+    conservationlaws(rn)
+    r = nps.rank
+    ig = incidencematgraph(rn)
+    lc = linkageclasses(rn)
+    nps.deficiency = Graphs.nv(ig) - length(lc) - r
+    nps.deficiency
+end
+
+# Used in the subsequent function.
+function subnetworkmapping(linkageclass, allrxs, complextorxsmap, p)
+    rxinds = sort!(collect(Set(rxidx for rcidx in linkageclass
+                               for rxidx in complextorxsmap[rcidx])))
+    rxs = allrxs[rxinds]
+    specset = Set(s for rx in rxs for s in rx.substrates if !isconstant(s))
+    for rx in rxs
+        for product in rx.products
+            !isconstant(product) && push!(specset, product)
+        end
+    end
+    specs = collect(specset)
+    newps = Vector{eltype(p)}()
+    for rx in rxs
+        Symbolics.get_variables!(newps, rx.rate, p)
+    end
+    rxs, specs, newps   # reactions and species involved in reactions of subnetwork
+end
```

Comment:
This was the kind of situation I mean when I said that I thought having subfunctions first and main functions second was confusing. In a file like this where almost everything has docstrings, when I look through it (on a first glance) looks like this function is grouped with the previous one (`deficiency`), and not with the subsequent one (`subnetworks`). In other situations I don't mind as much.